sharkfuser/include/fusili/logging.h

```diff
@@ -52,6 +54,141 @@ struct [[nodiscard]] ErrorObject {
+// preventing T from being created.
+//
+// usage:
+//   ErrorOr<AST> buildAST() {
+//      ErrorOr<std::string> maybeBuffer = getBuffer();
+//      FUSILI_CHECK_ERROR(maybeBuffer);
+//      AST ast = buildAST(*maybeBuffer);
+//      return ok(ast);
+//   }
+template <typename T> class [[nodiscard]] ErrorOr {
+  using Storage = std::variant<T, ErrorObject>;
+
+  Storage storage;
+
+  // The intended consumption pattern is to use `isOk` and `isError` utility
+  // methods to check this class when converted to an ErrorObject.
+  bool has_value() const noexcept { return std::holds_alternative<T>(storage); }
+
+  // Friend declaration to allow ErrorOr<U> to access ErrorOr<T>'s private
+  // members
+  template <typename U> friend class ErrorOr;
+
+public:
+  // Construct successful case from anything T is constructable from.
+  template <typename U>
+    requires std::constructible_from<T, U &&>
+  ErrorOr(U &&val) : storage(std::in_place_type<T>, std::forward<U>(val)) {}
+
+  // Copy constructor, to allow for ErrorOr<const char *> ->
+  // ErrorOr<std::string> for example.
+  template <typename U>
+    requires std::is_constructible_v<T, U>
+  ErrorOr(const ErrorOr<U> &other) {
+    if (isOk(other)) {
+      storage = Storage(std::in_place_type<T>, *other);
+    } else {
+      storage = Storage(std::in_place_type<ErrorObject>,
+                        std::get<ErrorObject>(other.storage));
+    }
+  }
+
+  // Move constructor, to allow for ErrorOr<const char *> ->
+  // ErrorOr<std::string> for example.
+  template <typename U>
+    requires std::is_constructible_v<T, U>
+  ErrorOr(ErrorOr<U> &&other) {
+    if (isOk(other)) {
+      storage = Storage(std::in_place_type<T>, std::move(*other));
+    } else {
+      storage = Storage(std::in_place_type<ErrorObject>,
+                        std::move(std::get<ErrorObject>(other.storage)));
+    }
+  }
+
+  // Construct implicitly from ErrorObject to support returning `error(...)`
+  // i.e.
+  //   ErrorOr<string> method() {
+  //     std::string output;
+  //     if(int returnCode = runShellScript(&output)) {
+  //        return error(ErrorCode::ScriptFail, "shell script error");
+  //     }
+  //     return ok(output) ;
+  //   }
+  ErrorOr(ErrorObject errorObject)
+      : storage(std::in_place_type<ErrorObject>, errorObject) {
+    assert(isError(std::get<ErrorObject>(storage)) &&
+           "successful results should be constructed with T type");
+  }
+
+  // Delete assignment operators.
+  ErrorOr &operator=(const ErrorOr &) = delete;
+  ErrorOr &operator=(ErrorOr &&) = delete;
+  template <typename U> ErrorOr &operator=(const ErrorOr<U> &) = delete;
+  template <typename U> ErrorOr &operator=(ErrorOr<U> &&) = delete;
+
+  // Convert to error object.
+  operator ErrorObject() const {
+    if (std::holds_alternative<T>(storage)) {
+      return ok();
+    }
+    return std::get<ErrorObject>(storage);
+  }
+
+#define ACCESSOR_ERROR                                                         \
```

Comment:
Perhaps add docstring here for readability (I understand what this is doing but it still helps)